libyul/optimiser/Solver.h

```diff
@@ -0,0 +1,92 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#pragma once
+
+#include <libyul/AST.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+
+// because of instruction
+#include <libyul/backends/evm/EVMDialect.h>
+
+#include <libsmtutil/SolverInterface.h>
+
+#include <memory>
+
+namespace solidity::smtutil
+{
+
+class SolverInterface;
+class Expression;
+struct Sort;
+
+}
+
+namespace solidity::yul
+{
+
+/**
+ * Base class for SMT based optimization steps.
+ *
+ * Works best when used with AST in SSA form.
+ */
+class Solver
```

Comment:
I would still suggest calling it SMTSolver or something else, as Solver might be confused with the LPSolver or other stuff.